system/jlib/jtrace.cpp

```diff
@@ -107,6 +108,293 @@ class NoopSpanExporterFactory
+            return "Producer";
+        case opentelemetry::trace::SpanKind::kConsumer:
+            return "Consumer";
+        case opentelemetry::trace::SpanKind::kInternal:
+            return "Internal";
+        default:
+            return "Unknown";
+    }
+}
+
+enum JLogExportLevel
+{
+    JLogExportLevelNone = 0,
+    JLogExportLevelMin = 1,
+    JLogExportLevelDefault = 2,
+    JLogExportLevelMax = 3
+};
+
+class JLogSpanExporter final : public opentelemetry::sdk::trace::SpanExporter
+{
+private:
+    JLogExportLevel logLevel = JLogExportLevelDefault;
+public:
+
+    inline JLogExportLevel jlogLevelFromStr(char const * str)
+    {
+        if (!isEmptyString(str))
+        {
+            if(strnicmp(str, "NONE", 4)==0)
+                return JLogExportLevelNone;
+            if(strnicmp(str, "MIN", 3)==0)
+                return JLogExportLevelMin;
+            if(strnicmp(str, "DEFAULT", 7)==0)
+                return JLogExportLevelDefault;
+            if(strnicmp(str, "MAX", 3)==0)
+                return JLogExportLevelMax;
+        }
+
+        return JLogExportLevelDefault;
+    }
+    JLogSpanExporter() {}
+
+    JLogSpanExporter(const char * logLevel)
+    {
+        setLogLevel(jlogLevelFromStr(logLevel));
+    }
+    
+    JLogSpanExporter(JLogExportLevel logLevel)
+    {
+        setLogLevel(logLevel);
+    }
+
+    void setLogLevel(JLogExportLevel level)
+    {
+        logLevel = level;
+    }
+
+    /**
+    * @return Returns a unique pointer to an empty recordable object
+    */
+    std::unique_ptr<opentelemetry::sdk::trace::Recordable> MakeRecordable() noexcept override
+    {
+        return std::unique_ptr<opentelemetry::sdk::trace::Recordable>(new opentelemetry::sdk::trace::SpanData());
+    }
+
+    /**
+    * Export - Formats recordable spans in HPCC Jlog format and reports to JLog
+    *
+    * @param recordables
+    * @return Always returns success
+    */
+    opentelemetry::sdk::common::ExportResult Export(
+      const nostd::span<std::unique_ptr<opentelemetry::sdk::trace::Recordable>> &recordables) noexcept override
+    {
+        if (isShutDown())
+        {
+            ERRLOG("JLog Trace Exporter: Export failed, exporter is shutdown");
+            return opentelemetry::sdk::common::ExportResult::kFailure;
+        }
+
+        for (auto &recordable : recordables)
+        {
+            auto span = std::unique_ptr<opentelemetry::sdk::trace::SpanData>(
+             static_cast<opentelemetry::sdk::trace::SpanData *>(recordable.release()));
+
+            if (span != nullptr)
+            {
+                char trace_id[32]       = {0};
+                char span_id[16]        = {0};
+                char parent_span_id[16] = {0};
+
+                span->GetTraceId().ToLowerBase16(trace_id);
+                span->GetSpanId().ToLowerBase16(span_id);
+                span->GetParentSpanId().ToLowerBase16(parent_span_id);
+
+                sout << "{"
```

Comment:
It is unusual to use stream syntax to create strings.  If there is a good reason then it is ok, but otherwise this is hard to read.  Also, the indentation is unhelpful (subsequent << should be indented).